File: cpp/2512. 奖励最顶尖的 K 名学生.cpp
class Solution {
public:
    vector<int> topStudents(vector<string>& positive_feedback, vector<string>& negative_feedback, vector<string>& report, vector<int>& student_id, int k) {
        unordered_map<string,int> words;
        for(auto p:positive_feedback){
            words[p]=3;
        }
        for(auto n:negative_feedback){
            words[n]=-1;
        }
        int len=student_id.size();
        vector<pair<int,int> > score_id;
        for(int i=0;i<len;i++){
            string rep=report[i];
            int stu_id=student_id[i];
            vector<string> wo=split(rep,' ');
            int score=0;
            for(auto w:wo){
                if(words.count(w)){
                    score+=words[w];
                }
            }
            score_id.push_back({-score,stu_id});
        }
        sort(score_id.begin(),score_id.end());
        vector<int> top_k;
        for(int i=0;i<k;i++){
            top_k.push_back(score_id[i].second);
        }
        return top_k;
    }

    vector<string> split(string& s, char c){
        vector<string> words;
        string word;
        for(auto str: s){
            if(str==c){
                words.push_back(word);
                word="";
                continue;
            }
            word.push_back(str);
        }
        words.push_back(word);
        return words;
    }
};

File: 刷题/golang/2512. 奖励最顶尖的 K 名学生.go
func topStudents(positive_feedback []string, negative_feedback []string, report []string, student_id []int, k int) []int {
    words := map[string]int{}
    for _, word := range positive_feedback{
        words[word]=3
    }
    for _, word := range negative_feedback {
        words[word]=-1
    }
    type pair struct { score, id int}
    var score_stuid []pair
    for i, rep := range report {
        score := 0
        for _, w := range strings.Split(rep, " ") {
            score += words[w]
        }
        score_stuid=append(score_stuid, pair{score, student_id[i]})
    }
    sort.Slice(score_stuid, func(i, j int) bool {
        a, b := score_stuid[i], score_stuid[j]
        return a.score > b.score || a.score == b.score && a.id<b.id
    })
    top_k:=make([]int,k)
    for i, p:=range score_stuid[:k] {
        top_k[i]=p.id
    }
    return top_k
}